usr/man/man8/chpt.8
.TH chpt 8 
.SH NAME
chpt \- change a disk partition table
.SH SYNTAX
.B /etc/chpt
[ \fB\-a\fR ]
[ \fB\-d\fR ]
[ \fB\-q\fR ]
[ \fB\-v\fR ]
[ [ \fB\-p\fR\fIx\fR \fIoffset\fR \fIsize\fR ] ... ]
.I device
.SH DESCRIPTION
The
.B chpt
command lets you alter the partition sizes
of a disk pack.  Using 
.B chpt, 
you can tailor your system disks and their
partitions to suit your system's individual needs.
.sp
If you want to create a file system on a partition that has been
modified, you must use newfs(8).
.PP
The standard procedure to change a partition table is:
.IP "1." 0.3i
Look at the current partition table using the \fB-q\fR
option.
.IP "2." 0.3i
If a file system does not exist on the \fIa\fR partition, create
one using the newfs(8) command.
.IP "" 0.3i
If a file system exists on the \fIa\fR partition but does
not contain a partition table in its superblock, copy the
partition table from the driver to the superblock using
the
.B chpt
command
with the \fB-a\fR option.
.IP "3." 0.3i
Change the partition offsets and sizes using the \fB-p\fIx\fR
option.  You can change all the partitions for one pack on one
command line.
.PP
The device
must be either the \fIa\fR or \fIc\fR partition of the raw
device, depending upon where the file system resides.  For
example, if the file system resides in the \fIa\fR
partition of an RM05 in drive 0, \fIdevice\fR is
rhp0a.
.PP
A file system must exist on the \fIa\fR or \fIc\fR partition
of the pack.  If you do not have a file system there, create one
using newfs(8).
.SH OPTIONS
.IP \fB\-a\fR 6
Copies the partition table in the device driver
to the disk pack.
.IP \fB\-d\fR 6
Copies the default partition table
to the disk pack and to the current partition table in the driver.
The default partition table is the table that was built with the 
disk driver.
.IP \fB\-q\fR
Runs
.PN chpt
without modifying the partition tables.
This prints the partition table of the specified disk pack. 
It prints the default partition table in the driver if there
is no partition table on the disk pack.
.IP \fB\-v\fR
Prints verbose messages showing
the progress of 
.PN chpt .
.IP \fB\-p\fR\fIx\fR
Changes the parameters of partition \fIx\fR on the disk pack
to the specified \fIoffset\fR and \fIsize\fR.
\fIx\fR is the partition you are modifying (a, b, c, d, e, f, g, or h).
\fIOffset\fR is the new beginning sector,
and \fIsize\fR is the new total number of sectors of the partition
being modified.
.SH EXAMPLES
This example shows how to change the partition table on an RM05
disk pack in drive 1.
The commands in this example change the the size of the \fIh\fR
partition to include the \fIg\fR partition.
Comments are in parenthesis to the right of commands.
.EX 0
.ta 2.5i
% chpt -q /dev/rhp1a	(\fRview partition table\fP)
/dev/rhp1a
No partition table found in superblock...
using default table from device driver.
Current partition table:
partition       bottom  top     size    overlap
    a                0   15883   15884  c
    b            16416   49855   33440  c
    c                0  500383  500384  a,b,d,e,f,g,h
    d           341696  357579   15884  c,g
    e           358112  414047   55936  c,g
    f           414048  500287   86240  c,g
    g           341696  500287  158592  c,d,e,f
    h            49856  341201  291346  c
%
.EE
.NT
In all of the tables generated by 
.B chpt,
\fIbottom\fR is the offset (starting sector),
\fItop\fR is the ending sector,
and \fIsize\fR is the number of sectors in the partition.
The \fIoverlap\fR is the other sectors that are partially
or entirely included in the partition.
.NE
.EX 0
.ta 2i
% bc	(\fRbasic calculator\fP)
500287-49856	(\fRtop of \fIg\fR minus bottom of \fIh\fR\f(CW)
450431
450431+1	(\fRadd 1 because it is zero\-based\fP)
450432	(\fRsize of new \fIh\fR partition\f(CW)
%
.EE
.PP
From the query,
you can see that there is no partition table in the superblock
of the \fIa\fR partition.
If this is because there is no file system in the \fIa\fR
partition,
run the newfs(8) command to create one.
.PP
For this example,
assume that there is a file system in the \fIa\fR partition of the disk,
but the file system does not contain a partition table in its
superblock.
Therefore, run 
.B chpt
with the \fB-a\fR option to copy the partition table
in the driver to the superblock of the \fIa\fR partition.
.EX 0
.ta 2.5i
% chpt -a /dev/rhp1a	(\fRadd table to \fIa\fR partition\f(CW)
%
.EE
.PP
Now you have a partition table to change.
.EX 0
.ta 4i
% chpt -v -ph 49856 450432 /dev/rhp1a	(\fRchange \fIh\fR\f(CW)
/dev/rhp1a
New partition table:
partition       bottom  top     size    overlap
    a                0   15883   15884  c
    b            16416   49855   33440  c
    c                0  500383  500384  a,b,d,e,f,g,h
    d           341696  357579   15884  c,g,h
    e           358112  414047   55936  c,g,h
    f           414048  500287   86240  c,g,h
    g           341696  500287  158592  c,d,e,f,h
    h            49856  500287  450432  c,d,e,f,g
%
.EE
.LP
.SH CAUTION
Changing partition tables indiscriminately
can result in losing large amounts of data.
.sp
Check for file systems on all the partitions of the disk
before using the \fB\-p\fR option.
If a file system exists whose partition may be destroyed,
copy it to a backup medium.
After you have changed the partitions,
restore the backed up file system.
.SH RESTRICTIONS
You must have superuser privileges to use
.B chpt.
.br
You can not shrink or change the offset of a partition
with a file system mounted on it
or with an open file descriptor on the entire partition.
.br
You can not change the offset of the \fIa\fR partition.
.SH "SEE ALSO"
ioctl(2), disktab(5), fsck(8), mkfs(8), newfs(8)


etc/chpt.c
#ifndef lint
static char sccsid[] = "@(#)chpt.c	1.8	(ultrix)	12/16/86";
#endif
/************************************************************************
 *									*
 *			Copyright (c) 1985, 1986 by			*
 *		Digital Equipment Corporation, Maynard, MA		*
 *			All rights reserved.				*
 *									*
 *   This software is furnished under a license and may be used and	*
 *   copied  only  in accordance with the terms of such license and	*
 *   with the  inclusion  of  the  above  copyright  notice.   This	*
 *   software  or  any  other copies thereof may not be provided or	*
 *   otherwise made available to any other person.  No title to and	*
 *   ownership of the software is hereby transferred.			*
 *									*
 *   The information in this software is subject to change  without	*
 *   notice  and should not be construed as a commitment by Digital	*
 *   Equipment Corporation.						*
 *									*
 *   Digital assumes no responsibility for the use  or  reliability	*
 *   of its software on equipment which is not supplied by Digital.	*
 *									*
 ************************************************************************/

/************************************************************************
 *
 *			Modification History
 *
 *	Paul Shaughnessy, 16-Dec-86
 * 005-	Added code that will open the device for read-only if the
 *	-q flag is specified.
 *
 *	Stephen Reilly, 24-Sep-85
 * 004- Added the -d switch, which will allow the user to set the disk
 *	partition tables back to the orginal size.
 *
 *	Stephen Reilly, 16-Apr-85
 * 003- Occasionally, when using the -q switch on a disk that does not
 *	have a valid superblock, the program will core dump.
 *	
 *	Stephen Reilly, 28-Jan-85
 * 002- Have the -q flag print out the partition table even if no superblock
 *	is present.
 *
 *	Stephen Reilly, 16-Jan-85
 * 001- Have -q flag print out the driver's table if it can't find the
 *	partition table in the superblock.
 *
 *	Mike Gancarz, 22-Nov-84
 * 000- Original version
 *
 ***********************************************************************/

/*
 * chpt	- change the partition tables on a disk
 *
 * usage: chpt [-a] [-q] [-v] [-d] [-px block-offset #blocks] device
 *  where:
 *	-a	Add partition tables to the disk if none are found
 *	-q	(query) run 'chpt' without writing new partition tables
 *		to the disk or setting them in the driver;implies -v
 *	-d	Add the paritition table to the disk with the default
 *		partition tables.
 *	-v	Print verbose messages showing what 'chpt' is doing
 *	-px	Override the driver's current partition values for
 *		partition 'x' with the specified block-offset and #blocks
 */

#include <stdio.h>
#include <sys/param.h>
#include <sys/fs.h>
#include <sys/ioctl.h>
#include <sys/file.h>
#include <sys/stat.h>

union {
	struct fs fs;
	char pad[MAXBSIZE];
} fsun;
#define	sblock	fsun.fs		/* used to access the s.b. as a file system */
#define sarray	fsun.pad	/* used to access the s.b. as a char array */

struct stat	st;		/* file status table */
struct pt	*newpt;		/* the new partition table to be written out */
struct
{
	int ovrlay[8];	/* these are set if part. was specified on cmd line */
	struct pt pt; /* partition values specified on the command line */
} av_pt;
#define av_part		av_pt.pt.pt_part[p]
#define new_part	newpt->pt_part[p]


char device[MAXPATHLEN];	/* full path to the special file */
char *part;			/* pointer to the device partition name */

int addpt;	/* if > 0, add part. tables to disk if none are found */
int query;	/* query mode */
int verbose;	/* verbose mode */
int dfpt;       /* 004 -d switch indicator */
int nopt;	/* if > 0, no partition table found in superblock */
int fd;		/* file descriptor */
int getcmd;	/* 004 type of ioctl request to use */

main(argc, argv)
int argc;
char *argv[];
{
	register int p;
	int q;
	register long i;
	long blk;

	/*
	 * must be superuser to run 'chpt'
	 */
	if (geteuid() != 0)
		fatal("must be superuser");

	/*
	 * process the argument list
	 */
	argc--, argv++;
	if (argc < 1)
		usage();

	while (argc > 0 && argv[0][0] == '-')
	{
		switch (argv[0][1])
		{
			case 'a':
				addpt++;
				break;

			case 'p':
				if (argv[0][2] < 'a' || argv[0][2] > 'h')
					fatal("bad switch: %s", *argv);
				p = argv[0][2] - 'a';

				--argc, ++argv;
				if (argc < 2)
					usage();
				if ((q = sscanf(*argv, "%d",
						&av_part.pi_blkoff)) != 1)
					usage();
				if (av_part.pi_blkoff < 0)
					fatal("negative block offset: %s",
						*argv);

				--argc, ++argv;
				if ((q = sscanf(*argv, "%ld",
						&av_part.pi_nblocks)) != 1)
					usage();
				if (av_part.pi_nblocks < 0)
					fatal("negative block count: %s",
						*argv);
				av_pt.ovrlay[p]++;
				break;

			case 'q':
				query++;
				/* FALLTHROUGH to case 'v' */

			case 'v':
				verbose++;
				break;

			case 'd':			/* 004 */
				dfpt++;
				break;
			default:
				fatal("bad switch: %s", *argv);
		}

		--argc, ++argv;
	}

	/*
	 * at this point, *argv points to the device (we hope)
	 */
	if (argc != 1)
		usage();
	strcpy(device, *argv);
	part = device + strlen(device) - 1;
	if ((*part != 'a') && (*part != 'c'))
		fatal("must specify `a' or `c' partition");

	/*
	 *	Which of the ioctl request will be used through the rest
	 *	of the program.
	 */
	if ( dfpt )
		getcmd = DIOCDGTPT;
	else
		getcmd = DIOCGETPT;
	/*
	 * let's be a pain and only allow the raw device
	 */
	if (stat(device, &st) < 0)
	{
		fprintf(stderr, "chpt: ");
		perror(device);
		exit(1);
	}
	if ((st.st_mode & S_IFMT) != S_IFCHR)
		fatal("%s: not a character device", device);

	/*
	 * read in the superblock from the named partition;
	 * complain if none exists
	 */
	if (verbose)
		printf("%s\n", device);
	/*
	 * 005 - If -q flag was specified, open device read-only.
	 */
	if (query)
		fd = open(device, O_RDONLY);
	else
		fd = open(device, O_RDWR);
	if (fd < 0)
	{
		fprintf(stderr, "chpt: cannot open %s", device);
		perror(" ");
		exit(1);
	}
	doseek(SBLOCK);
	if (read(fd, (char *)&sblock, SBSIZE) != SBSIZE)
	{
		fprintf(stderr, "chpt: %s: cannot read superblock", device);
		perror(" ");
		exit(1);
	}

	/*
	 * validate the superblock just read in
	 */
	if (sblock.fs_magic != FS_MAGIC && !query)
	       fatal("%s: blk %ld: bad superblock -- run fsck", device, SBLOCK);

	/*
	 * see if the superblock already has a partition table
	 */
	newpt = (struct pt *)&sarray[SBSIZE - sizeof(struct pt)];
	if (newpt->pt_magic != PT_MAGIC)
	{
		if (!addpt && !dfpt && !query)
			fatal("%s: no partition table in superblock", device);

		nopt++;
		if (verbose)
			printf("%s\n%s\n",
				"No partition table found in superblock...",
				"using default table from device driver.");
	}

	/*
	 * if the superblock has no partition table or are we using the
	 * the default table built into the driver ...
	 */
	if (nopt || dfpt)
	{
		/*
		 * ...then obtain a partition table from the driver and...
		 */
		if (ioctl(fd, getcmd, (char *)newpt) < 0)
		{
			perror("chpt: can't get partition table from driver");
			exit(1);
		}

		/*
		 * If the suprerblock is bad then skip the following
		 * code because we can core dump because of the divide
		 * instruction.
		 */
		if ( sblock.fs_magic != FS_MAGIC )		/* 001 */
			goto out;				/* 001 */

		/*
		 * ...warn the user if a new partition table would
		 * overwrite the rotational delay table in the driver
		 */
		sblock.fs_sbsize = SBSIZE; /* part. tbl. now part of size */
		blk = sblock.fs_spc * sblock.fs_cpc / NSPF(&sblock);
		for (i = 0; i < blk; i += sblock.fs_frag)
			/* void */;
		if ((struct pt *)(&sblock.fs_rotbl[(i - sblock.fs_frag) /
			sblock.fs_frag]) >= newpt)
		{
			printf("%s\n%s\n%s\n",
			  "Warning: new partition table overwriting existing",
			  "         rotational delay table. Filesystem",
			  "         performance may be impaired.");
			sblock.fs_cpc = 0;
		}
	}
out:

	/*
	 * overlay the partition table in the superblock with any values
	 * that the user may have specified on the command line; show
	 * what the new partition table looks like
	 */
	ptoverlay();
	if (verbose)
		report();

	/*
	 * set the new partition table in the driver via ioctl(2)
	 * and write the updated superblock back out to disk
	 */
	newpt->pt_magic = PT_MAGIC;
	newpt->pt_valid = PT_VALID;
	if (!query)
	{
		if (ioctl(fd, DIOCSETPT, (char *)newpt) < 0)
		{
			perror("chpt: cannot set partition table in driver");
			exit(1);
		}
		writefs();
	}
	exit(0);
}

/*
 * overlay the partition table in the superblock
 */
ptoverlay()
{
	register int p;

	for (p = 0; p < 8; p++)
	{
		if (av_pt.ovrlay[p])
		{
			new_part.pi_nblocks = av_part.pi_nblocks;
			new_part.pi_blkoff = av_part.pi_blkoff;
		}
	}
}

/*
 * report on overlapping partitions in the new partition table, regardless
 * of their mount status
 */
report()
{
	register int p, np;
	register daddr_t bot[8], top[8];
	int flag;

	for (p = 0; p < 8; p++)
	{
		bot[p] = new_part.pi_blkoff;
		top[p] = bot[p] + new_part.pi_nblocks;
		if (top[p] >= 2)
			top[p]--;
	}
	printf("%s partition table:\n", query ? "Current" : "New");
	printf("%-10s%10s %10s %10s %s\n",
		"partition", "bottom", "top", "size", "   overlap");
	for (p = 0; p < 8; p++)
	{
		printf("    %c     %10ld %10ld %10ld    ",
			'a' + p, bot[p], top[p], new_part.pi_nblocks);
		flag = 0;
		for (np = 0; np < 8; np++)
		{
			if (p == np)
				continue;
			if ((bot[p] >= bot[np] && bot[p] <= top[np]) ||
			    (top[p] >= bot[np] && top[p] <= top[np]) ||
			    (bot[np] >= bot[p] && bot[np] <= top[p]))
			{
				printf("%s%c", flag ? "," : "", 'a' + np);
				flag++;
			}
		}
		putchar('\n');
	}
}

/*
 * re-write the superblocks on the disk, trashing each one as we go
 * (jes' kiddin') 
 */
writefs()
{
	long cylno;

	wrblk(SBLOCK, SBSIZE, (char *)&sblock);

	for (cylno = 0; cylno < sblock.fs_ncg; cylno++)
		wrblk(fsbtodb(&sblock, cgsblock(&sblock, cylno)),
		      SBSIZE, (char *)&sblock);
}

/*
 * seek to your favorite block on the disk
 */
doseek(blkno)
daddr_t blkno;
{
	if (lseek(fd, blkno * DEV_BSIZE, 0) < 0)
	{
		fprintf(stderr, "chpt: seek error: %ld", blkno * DEV_BSIZE);
		perror(" ");
		exit(1);
	}
}

/*
 * write a block to the disk
 */
wrblk(blkno, size, buf)
daddr_t blkno;
int size;
char *buf;
{
	int n;

	doseek(blkno);
	n = write(fd, buf, size);
	if (n != size)
	{
		fprintf(stderr, "chpt: write error: %ld", blkno);
		perror(" ");
		exit(1);
	}
}

/*VARARGS*/
/*
 * print an error message on stderr and exit
 */
fatal(fmt, arg1, arg2)
char *fmt;
{
	fprintf(stderr, "chpt: ");
	fprintf(stderr, fmt, arg1, arg2);
	putc('\n', stderr);
	exit(1);
}

/*
 * print a "usage" message and exit
 */
usage()
{
	printf("usage: chpt [-a] [-q] [-v] [-d] [-px blk-offset size] device\n");
	exit(1);
}
